ENH: make guess_datetime_format public

nnlnr opened the pull request:
- [X] closes #54727 (Replace xxxx with the GitHub issue number)
- [X] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [X] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [NA] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [X] Added an entry in the latest `doc/source/whatsnew/v2.1.1.rst` file if fixing a bug or adding a new feature.

Commit:
updating init to include guess_datetime_format func

Commit:
included guess_datetime_format in api

Commit:
Merge remote-tracking branch 'upstream/main' into infer-datetime-54727

Commit:
updated test to include guess_datetime_format method

Commit:
ran pre-commit hook

Commit:
Merge remote-tracking branch 'upstream/main' into infer-datetime-54727

Commit:
updated latest rst doc with enhancement note

mroeschke commented on doc/source/whatsnew/v2.1.1.rst:
```diff
@@ -42,7 +42,7 @@ Bug fixes
 Other
 ~~~~~
--
+- Enhancement in :meth:`_libs.tslibs.parsing.guess_datetime_format` to make public (:issue:`54727`)
```
```suggestion
- :meth:`pandas.tseries.guess_datetime_format` is now available (:issue:`54727`)
```

Can you move this to the other enhancement section?

mroeschke replied:
Also this should be in 2.2.0.rst

nnlnr replied:
WIll do - thanks for the review!

nnlnr replied:
moved enhancement note to `2.2.0.rst` unde the **Other Enhancements** section ✅

mroeschke reviewed (changes requested):
Could you add this to the API reference documentation?

Commit:
Merge remote-tracking branch 'upstream/main' into infer-datetime-54727

Commit:
removed enhancement update

Commit:
alphabetize other-enhancement list

nnlnr commented:
> Could you add this to the API reference documentation?
is the API reference documentation in` doc/source/reference/api`? If so, I'm seeing that path in the .gitignore file - is there a procedure to update the API reference under this context?

mroeschke commented:
I think you can just add an entry to `doc/source/reference/general_functions.rst`

MarcoGorelli reviewed (changes requested):
thanks for your pr

whilst we're changing this

- in the `guess_Datetime_format` docstring, where it says

>         Warning: dayfirst=True is not strict, but will prefer to parse
        with day first (this is a known bug).

could you write that as a sphinx directive? so `.. warning::`. check the `to_datetime` docs, there's something similar there

- could we add an example to the docstring please? just a simple one-liner of it guessing a datetime format, and another one-liner of it not being able to guess another one, will do

Commit:
added guess_datetime_format

Commit:
updated guess_datetime_format docstring with warning sphinx directive

Commit:
fixed spelling from pre-commit hook

Commit:
Merge remote-tracking branch 'upstream/main' into infer-datetime-54727

nnlnr commented:
Thanks for the review and comments @mroeschke and @MarcoGorelli 
- added an entry to doc/source/reference/general_functions.rst ✅ 
- updated warning in docstring as sphinx directive ✅ 
- added examples to docstring ✅

MarcoGorelli commented:
Can you fix the failing ci checks please

Commit:
Merge remote-tracking branch 'upstream/main' into infer-datetime-54727

Commit:
update guess_datetime_format docstring examples

Commit:
fixed trailing whitespace

Commit:
correct sphinx directive for guess_datetime_format with trailing whitespace

nnlnr commented:
👋  I'm getting a bit stuck with these failing checks - which seem to be mostly due to the following attribute error:
`AttributeError: module 'pandas' has no attribute 'guess_datetime_format'`

I updated the docstring in parsing.pyx with a sphinx `..warning::` directive and added two examples (learned that formatting was important which was a good lesson); however not sure how to address the failing jobs - any feedback or advice on how to troubleshoot?

Commit:
Merge remote-tracking branch 'upstream/main' into infer-datetime-54727

Commit:
updated whatsnew with corrected guess_datatime_format api method string

mroeschke commented on doc/source/whatsnew/v2.1.1.rst:
```diff
@@ -43,7 +43,7 @@ Bug fixes
 
 Other
 ~~~~~
--
```
Can you undo this?

mroeschke commented on pandas/_libs/tslibs/parsing.pyx:
```diff
@@ -874,14 +874,28 @@ def guess_datetime_format(dt_str: str, bint dayfirst=False) -> str | None:
+
+    Examples
+    --------
+    >>> # Successful guess
+    >>> # Returns a datetime formatted string
```
```suggestion
```

mroeschke commented on pandas/_libs/tslibs/parsing.pyx:
```diff
@@ -874,14 +874,28 @@ def guess_datetime_format(dt_str: str, bint dayfirst=False) -> str | None:
+    >>> guess_datetime_format('09/13/2023')
+    '%m/%d/%Y'
+
+    >>> # Unsuccessful guess due to date string with extra or unexpected characters
+    >>> # Returns nothing
```
```suggestion
```

mroeschke commented on doc/source/whatsnew/v2.2.0.rst:
```diff
@@ -45,8 +45,8 @@ enhancement2
 
 Other enhancements
 ^^^^^^^^^^^^^^^^^^
+- :meth:`pandas.tseries.api.guess_datetime_format` is now available (:issue:`54727`)
```
```suggestion
- :meth:`pandas.tseries.guess_datetime_format` is now available (:issue:`54727`)
```

nnlnr replied:
Is this because guess_datetime_format is now a public method - and no longer needs to be called through `api`?

mroeschke replied:
Oh sorry, I'm actually incorrect. I see you added it to `api`.

mroeschke commented on doc/source/reference/general_functions.rst:
```diff
@@ -58,6 +58,7 @@ Top-level dealing with datetimelike data
    period_range
    timedelta_range
    infer_freq
+   guess_datetime_format
```
For this entry specifically I think you'll need a new autosummary section with 

`.. currentmodule:: pandas.tseries.frequencies`

nnlnr replied:
Tried adding a new autosummary section with 
`.. currentmodule:: pandas.tseries.frequencies`
as follows:
```
Top-level dealing with datetimelike data
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.. autosummary::
   :toctree: api/


   to_datetime
   to_timedelta
   date_range
   bdate_range
   period_range
   timedelta_range
   infer_freq


.. autosummary::
   :currentmodule: pandas.tseries.frequencies


   guess_datetime_format
```

However, I'm still getting hit with the `"AttributeError: module 'pandas' has no attribute 'guess_datetime_format'"` for docstring validation - 

should this be a completely different section with a different header?

Commit:
removed extra line

Commit:
removed comment lines in example for guess_datetime_format

Commit:
removed api from method call for guess_datetime_format

Commit:
pre-commit hook fix for whitespace

Commit:
removed other guess_datetime_format

Commit:
fix extra char

Commit:
Merge remote-tracking branch 'upstream/main' into infer-datetime-54727

Commit:
fix whitespace

Commit:
removed toctree api

nnlnr commented:
👋 Marco - thanks! That would be great - let me fix up the conflicts and circle back here

Commit:
fixed conflicts

Commit:
Merge remote-tracking branch 'upstream/main' into infer-datetime-54727

MarcoGorelli committed:
fixup

MarcoGorelli committed:
Merge remote-tracking branch 'upstream/main' into pr/nnlnr/infer-datetime-54727

MarcoGorelli committed:
try fixing docstring

MarcoGorelli committed:
add whatsnew note

MarcoGorelli commented:
website build looks good:

![image](https://github.com/pandas-dev/pandas/assets/33491632/ce87e0b4-f2ac-417f-a5ba-e67c020370a1)

merging then, thanks @nnlnr !

MarcoGorelli reviewed (approved):
going to need @mroeschke to approve before this can be merged

MarcoGorelli commented on doc/source/reference/general_functions.rst:
```diff
@@ -61,6 +61,7 @@ Top-level dealing with datetimelike data
 
 Top-level dealing with Interval data
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
 .. autosummary::
    :toctree: api/
```
extra line, but it doesn't seem to make any difference in the build so is OK imo

![image](https://github.com/pandas-dev/pandas/assets/33491632/d9b301f8-1379-4ce6-8363-573f652f7915)

mroeschke commented on doc/source/reference/general_functions.rst:
```diff
@@ -61,6 +61,7 @@ Top-level dealing with datetimelike data
 
 Top-level dealing with Interval data
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
```
```suggestion
```

mroeschke commented:
Could you also add a whatsnew note in 2.2 under the `Other Enhancements` section?

mroeschke commented:
Thanks @nnlnr  and @MarcoGorelli

nnlnr commented:
Thanks all 🎉 🙏
